notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+    "# - \"edits\" represent the number of edits done by the user, so just count the number of events\n",
+    "# - calculate the registration time using the minimum __time from events and adjusting it some variable years back in time\n",
+    "# - determine the preferred language of the user based on their earliest channel edit\n",
+    "\n",
+    "sql = '''\n",
+    "REPLACE INTO \"wiki_users\" OVERWRITE ALL\n",
+    "SELECT \n",
+    "  \"user\", \n",
+    "  EARLIEST(\n",
+    "    CASE \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),5) > 3 THEN 'Reviewers' \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),17) > 13 THEN 'Patrollers' \n",
+    "      WHEN  MOD(TIMESTAMP_TO_MILLIS(__time),23) > 21 THEN 'Bots'\n",
+    "      ELSE 'Autoconfirmed'\n",
+    "    END,\n",
+    "    1024\n",
+    "  ) AS \"group\",\n",
+    "  count(*) \"edits\",\n",
+    "  TIME_SHIFT(MIN(__time), 'P1Y', -1 * MOD(MIN(EXTRACT (MICROSECOND FROM __time)),20) ) AS \"registered_at_ms\",\n",
+    "  EARLIEST(SUBSTRING(\"channel\", 2, 2), 1024) AS \"language\"\n",
+    "FROM \"wikipedia_events\"\n",
+    "GROUP BY 1\n",
+    "PARTITIONED BY ALL\n",
+    "'''\n",
+    "request = druid.sql.sql_request( sql)              # init request object\n",
+    "request.add_context( 'finalizeAggregations', True) # EARLIEST functions will store a partial aggregation otherwise\n",
+    "request.add_context( 'maxNumTasks', 2)             # can't go any higher in test env\n",
+    "\n",
+    "druid.display.run_task(request)\n",
+    "druid.sql.wait_until_ready('wiki_users')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0f3ff8bb-2a1d-4445-be59-9d2460397875",
+   "metadata": {},
+   "source": [
+    "Okay, now we are ready to try an ingestion using the sortMerge join:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d1bcc958-357c-4b64-b19f-bcbf0fc6941b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql = '''\n",
+    "REPLACE INTO \"wiki_merge\" OVERWRITE ALL\n",
+    "WITH \"wikidata\" AS \n",
+    "(\n",
+    "    SELECT *\n",
+    "    FROM TABLE(\n",
+    "      EXTERN(\n",
+    "        '{\"type\":\"http\",\n",
+    "          \"uris\":[ \"https://druid.apache.org/data/wikipedia.json.gz\"]\n",
+    "         }',\n",
+    "        '{\"type\":\"json\"}'\n",
+    "      )\n",
+    ") EXTEND (\"isRobot\" VARCHAR, \"channel\" VARCHAR, \"timestamp\" VARCHAR, \"flags\" VARCHAR, \"isUnpatrolled\" VARCHAR, \"page\" VARCHAR, \"diffUrl\" VARCHAR, \"added\" BIGINT, \"comment\" VARCHAR, \"commentLength\" BIGINT, \"isNew\" VARCHAR, \"isMinor\" VARCHAR, \"delta\" BIGINT, \"isAnonymous\" VARCHAR, \"user\" VARCHAR, \"deltaBucket\" BIGINT, \"deleted\" BIGINT, \"namespace\" VARCHAR, \"cityName\" VARCHAR, \"countryName\" VARCHAR, \"regionIsoCode\" VARCHAR, \"metroCode\" BIGINT, \"countryIsoCode\" VARCHAR, \"regionName\" VARCHAR))\n",
+    "SELECT \n",
```

Comment:
Do you need all these dimensions? As per @techdocsmith I'd suggest being specific about the columns here to show good practice.

Reply:
let's move on for now.... I get that it looks cleaner but I think 28 work is more important.